File: jsrc/legacy/SxxMachine/builtin/PRED_mutex_trylock_1.java
package SxxMachine;

import java.util.concurrent.locks.Lock;

import SxxMachine.FFIObjectTerm;
import SxxMachine.Mutex;
import SxxMachine.Operation;
import SxxMachine.Predicate.P1;
import SxxMachine.Prolog;
import SxxMachine.SymbolTerm;
import SxxMachine.Term;
import SxxMachine.IllegalTypeException;
import SxxMachine.PrologException;
/**

<p>mutex_trylock(+MutexId)
    As  mutex_lock/1,  but if the mutex  is held by another thread,  this
    predicates fails immediately.
    
  <p>description is copied from swi prolog documentation
 * @author semenov
 *
 */
public class PRED_mutex_trylock_1 extends P1 {

	public PRED_mutex_trylock_1(Term arg1, Operation cont) {
		this.LARG[0] = arg1;
		this.cont = cont;
	}

	@Override
	public Operation exec(Prolog engine) throws PrologException {
		Term a1 = LARG[0].dref();

		Lock lock;
		if ((a1 instanceof SymbolTerm)){
			lock = Mutex.getInstance(a1.name());
		} else if ((a1 instanceof FFIObjectTerm) && (a1.toJava() instanceof Lock)) {
			lock = (Lock) a1.toJava();
		} else {
			throw new IllegalTypeException(this, 1, "atom or FFIObjectTerm(Lock)", a1);
		}
		if (!lock.tryLock()){
			return engine.fail();
		}
		return cont;
	}
}

File: src/legacy/SxxMachine/builtin/PRED_mutex_trylock_1.cpp
using namespace std;

#include "PRED_mutex_trylock_1.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../exceptions/SxxMachine/PrologException.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "../../../machine/SxxMachine/Mutex.h"
#include "../../../machine/SxxMachine/FFIObjectTerm.h"
#include "../../../exceptions/SxxMachine/IllegalTypeException.h"

namespace SxxMachine
{
	using FFIObjectTerm = SxxMachine::FFIObjectTerm;
	using Mutex = SxxMachine::Mutex;
	using Operation = SxxMachine::Operation;
	using P1 = SxxMachine::Predicate::P1;
	using Prolog = SxxMachine::Prolog;
	using SymbolTerm = SxxMachine::SymbolTerm;
	using Term = SxxMachine::Term;
	using IllegalTypeException = SxxMachine::IllegalTypeException;
	using PrologException = SxxMachine::PrologException;

	PRED_mutex_trylock_1::PRED_mutex_trylock_1(Term *arg1, Operation cont)
	{
		this->LARG[0] = arg1;
		this->cont = cont;
	}

	Operation PRED_mutex_trylock_1::exec(Prolog *engine) throw(PrologException)
	{
		Term *a1 = LARG[0]->dref();

		Lock *lock;
		if ((dynamic_cast<SymbolTerm*>(a1) != nullptr))
		{
			lock = Mutex::getInstance(a1->name());
		}
		else if ((dynamic_cast<FFIObjectTerm*>(a1) != nullptr) && (dynamic_cast<Lock*>(a1->toJava()) != nullptr))
		{
			lock = any_cast<Lock*>(a1->toJava());
		}
		else
		{
			throw IllegalTypeException(this, 1, "atom or FFIObjectTerm(Lock)", a1);
		}
		if (!lock->tryLock())
		{
			return engine->fail();
		}
		return cont;
	}
}
